File: src/main/java/io/github/arieslab/base/cobertura/ReportGenerator.java
package io.github.arieslab.base.cobertura;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

import org.jacoco.core.analysis.Analyzer;
import org.jacoco.core.analysis.CoverageBuilder;
import org.jacoco.core.analysis.IBundleCoverage;
import org.jacoco.core.tools.ExecFileLoader;
import org.jacoco.report.DirectorySourceFileLocator;
import org.jacoco.report.IReportVisitor;
import org.jacoco.report.csv.CSVFormatter;

public class ReportGenerator {

	private static final Logger LOGGER = Logger.getLogger(ReportGenerator.class.getName());

	private final String title;

	private File executionDataFile;
	private File classesDirectory;
	private File sourceDirectory;
	private File reportDirectory;

	private ExecFileLoader execFileLoader;

	/**
	 * Creates a new ReportGenerator for the given project directory with the specified coverage report output directory.
	 *
	 * @param projectDirectory the project root directory
	 * @param coveragereport the directory where the coverage report will be written
	 */
	public ReportGenerator(final File projectDirectory, final File coveragereport) {
		this.title = projectDirectory.getName();
		this.executionDataFile = new File(projectDirectory, "target/jacoco.exec");
		this.classesDirectory = new File(projectDirectory, "target/generated-classes");

		if(!this.classesDirectory.exists()){
			this.classesDirectory = new File(projectDirectory, "target/classes");
		}

		this.sourceDirectory = new File(projectDirectory, "src/main/java");
		this.reportDirectory = new File(coveragereport, "");
	}

	/**
	 * Loads execution data, analyzes the structure, and creates the CSV report.
	 */
	public void create() throws IOException {
		loadExecutionData();
		final IBundleCoverage bundleCoverage = analyzeStructure();
		createReport(bundleCoverage);
	}

	/**
	 * Generates a CSV coverage report from the analyzed bundle structure.
	 *
	 * @param bundleCoverage the coverage data bundle
	 */
	private void createReport(final IBundleCoverage bundleCoverage)
			throws IOException {
		final CSVFormatter csvFormatter = new CSVFormatter();
		reportDirectory.mkdir();
		IReportVisitor visitor = csvFormatter.createVisitor(new FileOutputStream(new File(reportDirectory, this.title+"_jacoco.csv")));

		visitor.visitInfo(execFileLoader.getSessionInfoStore().getInfos(),
				execFileLoader.getExecutionDataStore().getContents());

		visitor.visitBundle(bundleCoverage, new DirectorySourceFileLocator(
				sourceDirectory, "utf-8", 4));

		visitor.visitEnd();
	}

	/**
	 * Loads JaCoCo execution data from the jacoco.exec file.
	 */
	private void loadExecutionData() throws IOException {
		execFileLoader = new ExecFileLoader();
		execFileLoader.load(executionDataFile);
	}

	/**
	 * Analyzes all compiled classes to build a coverage bundle.
	 *
	 * @return the bundle coverage data
	 */
	private IBundleCoverage analyzeStructure() throws IOException {
		final CoverageBuilder coverageBuilder = new CoverageBuilder();
		final Analyzer analyzer = new Analyzer(
				execFileLoader.getExecutionDataStore(), coverageBuilder);

		try {
			analyzer.analyzeAll(classesDirectory);
		}catch (Exception e){
			LOGGER.log(Level.WARNING, "Failed to analyze classes directory", e);
		}

		return coverageBuilder.getBundle(title);
	}


}

File: src/main/java/io/github/arieslab/base/cobertura/ReportGenerator.kt
package io.github.arieslab.base.cobertura

import org.jacoco.core.analysis.Analyzer
import org.jacoco.core.analysis.CoverageBuilder
import org.jacoco.core.analysis.IBundleCoverage
import org.jacoco.core.tools.ExecFileLoader
import org.jacoco.report.DirectorySourceFileLocator
import org.jacoco.report.csv.CSVFormatter
import java.io.File
import java.io.FileOutputStream
import java.util.logging.Logger

class ReportGenerator(
    projectDirectory: File,
    coveragereport: File,
    type: BuildType = BuildExecutor.detect(projectDirectory)
) {
    private val LOGGER = Logger.getLogger(ReportGenerator::class.java.name)

    private val title: String = projectDirectory.name
    private val executionDataFile: File = type.getJacocoExecFile(projectDirectory)
    private var classesDirectory: File = type.getClassesDir(projectDirectory).takeIf { it.exists() }
        ?: File(projectDirectory, "target/classes")
    private var sourceDirectory: File = type.getSourceDir(projectDirectory).takeIf { it.exists() }
        ?: File(projectDirectory, "src/main/java")
    private val reportDirectory: File = File(coveragereport, "")
    private val execFileLoader = ExecFileLoader()

    @Throws(java.io.IOException::class)
    fun create() {
        loadExecutionData()
        val bundleCoverage = analyzeStructure()
        createReport(bundleCoverage)
    }

    @Throws(java.io.IOException::class)
    private fun createReport(bundleCoverage: IBundleCoverage) {
        val csvFormatter = CSVFormatter()
        reportDirectory.mkdir()
        val csvFile = File(reportDirectory, "${title}_jacoco.csv")
        val visitor = csvFormatter.createVisitor(FileOutputStream(csvFile))
        visitor.visitInfo(execFileLoader.sessionInfoStore.infos, execFileLoader.executionDataStore.contents)
        visitor.visitBundle(bundleCoverage, DirectorySourceFileLocator(sourceDirectory, "utf-8", 4))
        visitor.visitEnd()
    }

    @Throws(java.io.IOException::class)
    private fun loadExecutionData() {
        execFileLoader.load(executionDataFile)
    }

    @Throws(java.io.IOException::class)
    private fun analyzeStructure(): IBundleCoverage {
        val coverageBuilder = CoverageBuilder()
        val analyzer = Analyzer(execFileLoader.executionDataStore, coverageBuilder)
        try {
            analyzer.analyzeAll(classesDirectory)
        } catch (e: Exception) {
            LOGGER.warning("Failed to analyze classes directory: $e")
        }
        return coverageBuilder.getBundle(title)
    }
}
